showtime-xyz/showtime-frontend
# refactor(SHOW2-435): Refactored Auth / User Info

# Why

Currently we use the `useUser` hook in multiple places to extract `isAuthenticated` or `user` object, however this hook has an effect that every-time the access token gets updated it will refresh the access token immediately.

Although, the `AccessToken` class will block redundant refresh calls, however the hook will trigger multiple unwanted state renders.

# How

https://user-images.githubusercontent.com/4061838/152250611-b2ce2433-317a-4f68-96d7-5daafb2b26f9.mov

- Extracted auth functionalities into multiple small pieces 
  - `AuthContext` & `AuthProvider` to provide authentication functionalities and variables, like:
    - `login` & `logout` functions
    - `accessToken` & `authenticationStatus` variables
  - `useMagicLogin` hook to provide login functionalities with `Magic` provider
  - `useWalletLogin` hook to provide login functionalities with `walletconnect` provider
  - `useAuth` hook to provide `AuthContext` variables
- Extracted user info functionalities into a separate context and hook
  - `UserContext` & `UserProvider` to provide fetching user profile functionality and variables, like: 
    - `user` & `error` objects
    - `isLoading` & `isAuthenticated` bools
  - `useUser` hook to provide `UserContext` variables
- Refactored `useLogin` to utilise the new created hooks

# Test Plan

### Test Phone Number Login
- make sure you clear the app data and cache
- launch the app
- tap on `sign in` on the header
- enter your phone number
- > app should present magic login ui
- enter your sms code
- > app should log you in

### Test Email Login
- make sure you clear the app data and cache
- launch the app
- tap on `sign in` on the header
- tap on `other options`
- enter your email address
- > app should present magic login ui
- go to you mail app, and click on the incoming email link
- web page will open and redirect you to the app
- > app should log you in

### Test Wallet Login
- make sure you clear the app data and cache
- launch the app
- tap on `sign in` on the header
- tap on `other options`
- tap on `sign in with wallet`
- > app should a modal to select your wallet
- > app should redirect you to the wallet app to connect
- > grant access to ShowTime app to access wallet
- > app should become active
- > app should present connecting to wallet loading message
- > app should redirect you to the wallet app to sign personal message
- > grant access to ShowTime to sign a message
- > app should become active 
- > app should log you in
- > app should dismiss the login modal
